XenAdmin/SettingsPanels/SnmpEditPage.cs

```diff
@@ -0,0 +1,278 @@
+                _invalidParamToolTip = new ToolTip
+                {
+                    IsBalloon = true,
+                    ToolTipIcon = ToolTipIcon.Warning,
+                };
+                _invalidParamToolTipText = " ";
+
+                if (!EnableSnmpCheckBox.Checked)
+                {
+                    return true;
+                }
+                if (!SupportV2cCheckBox.Checked && !SupportV3CheckBox.Checked)
+                {
+                    _invalidParamToolTip.Tag = SupportV2cCheckBox;
+                    _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_CHOOSE_TITLE;
+                    return false;
+                }
+                if (SupportV2cCheckBox.Checked)
+                {
+                    var communityStr = CommunityTextBox.Text;
+                    if (!RegexCommon.Match(communityStr.Trim()).Success)
+                    {
+                        _invalidParamToolTip.Tag = CommunityTextBox;
+                        _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_COMMUNITY_TITLE;
+                        _invalidParamToolTipText = Messages.SNMP_ALLOW_COMMUNITY_TEXT;
+                        return false;
+                    }
+                }
+                if (SupportV3CheckBox.Checked)
+                {
+                    var securityStr = UserNameTextBox.Text;
+                    var authPassStr = AuthenticationPasswordLabelTextBox.Text;
+                    var privacyPassStr = PrivacyPasswordTextBox.Text;
+                    if (!RegexCommon.Match(securityStr.Trim()).Success)
+                    {
+                        _invalidParamToolTip.Tag = UserNameTextBox;
+                        _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_USER_TITLE;
+                        _invalidParamToolTipText = Messages.SNMP_ALLOW_COMMUNITY_TEXT;
+                        return false;
+                    }
+
+                    if (!RegexEncryptTextBox.Match(authPassStr.Trim()).Success)
+                    {
+                        _invalidParamToolTip.Tag = AuthenticationPasswordLabelTextBox;
+                        _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_AUTH_TITLE;
+                        _invalidParamToolTipText = Messages.SNMP_ALLOW_AUTH_TEXT;
+                        return false;
+                    }
+
+                    if (!RegexEncryptTextBox.Match(privacyPassStr.Trim()).Success)
+                    {
+                        _invalidParamToolTip.Tag = PrivacyPasswordTextBox;
+                        _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_PRIVACY_TITLE;
+                        _invalidParamToolTipText = Messages.SNMP_ALLOW_AUTH_TEXT;
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+        public SnmpEditPage()
+        {
+            InitializeComponent();
+            Text = Messages.SNMP;

```

Comment:
Setting virtual members in constructors can cause problems in derived classes. It is better to override it:
`public override string Text => Messages.SNMP;`

(Note that the code elsewhere is inconsistent and you may see this in other files; these need fixing eventually; we'll do it as we go).